SVG: optimize and refactor group creation code

timvandermeij opened the pull request:
This patch series brings down the number of transform groups for the Tracemonkey paper from 2790 to 1271, resulting in a much lighter DOM and faster rendering/scrolling. Refer to the commit messages for more information.

timvandermeij committed:
Examples: improve SVG viewer

This patch:

- resolves a warning in the console about missing character encoding;
- makes the viewer use the same background color and PDF file as the
  regular viewer;
- simplifies the example to bring it more in line with the other
  examples.

timvandermeij committed:
SVG: optimize transform group creation

This patch ensures that we only create transformation groups when it is
actually required and that we re-use transform groups as much as possible.
It reduces the number of transform groups for the Tracemonkey paper from
2790 to 1271, thereby making the DOM much lighter and rendering/scrolling
smoother. Moreover, it simplifies the code and prevents duplication.

Finally, we issue a warning when an unimplemented graphic state is
encountered. Before, this was ignored silently, making debugging harder.

yurydelendik commented on src/display/svg.js:
```diff
@@ -464,14 +462,9 @@ var SVGGraphics = (function SVGGraphicsClosure() {
-        this.pgrp.appendChild(this.defs);
-        this.pgrp.appendChild(this.tgrp);
-        this.svg.appendChild(this.pgrp);
+        this.svg.setAttributeNS(null, 'transform', pm(viewport.transform));
```
this does not work for chrome -- we will need to create a root group and use it as this.svg ?

timvandermeij replied:
Fixed in the new commits.

yurydelendik commented on src/display/svg.js:
```diff
@@ -1200,8 +1158,30 @@ var SVGGraphics = (function SVGGraphicsClosure() {
+    /**
+     * @private
+     */
+    _finishGroup: function SVGGraphics_finishGroup() {
```
_finishGroup can be removed, see https://github.com/yurydelendik/pdf.js/commit/f4c8099678a0d19f1e48c891aa55c2581931495f

timvandermeij replied:
Fixed in the new commits. Nice find!

yurydelendik reviewed (changes requested):
That is a really good start. I think we need to backout some changes in the viewer and avoid a regression in chrome.

timvandermeij committed:
SVG: factor out initialization code into a private method

Each well-formed SVG image has the following structure:

SVG element
- Definitions element
- Root group
  - Other group 1
  - ...
  - Other group n

This patch factors out initialization code into a private method in such
as way that the creation of this structure is clear from the code. The
root group is the replacement for the parent group from before. We need
this group as we cannot apply the viewport transform on the SVG element
itself (this caused issues in Chrome). If other code appends groups to
the SVG image, in reality it is appending those groups to the root
group, but this detail is abstracted away by this patch.

timvandermeij commented:
I have updated the commits based on the review comments.

yurydelendik reviewed (approved):
Thank you for the patch.
